File: Download.fs
module comicdownloader.Download

open System.Net.Http
open System.Runtime.CompilerServices
open System.Security.Cryptography
open System
open System.Threading.Tasks
open SixLabors.ImageSharp
open SixLabors.ImageSharp.PixelFormats
open Microsoft.Extensions.Logging
open Comics

let private getFileName (bytes: byte array) =
    let extension =
        match bytes[0..4] with
        | [| 0xFFuy; 0xD8uy; _; _ |] -> ".jpg"
        | [| 0x42uy; 0x4Duy; _; _ |] -> ".bmp"
        | [| 0x47uy; 0x49uy; 0x46uy |] -> ".gif"
        | _ -> ".png"

    DateTime.Now.ToString("yyyy.MM.dd") + extension

let private getHash comicName (bytes: byte array) =
    let image = Image.Load<Rgba32>(bytes)

    let pixelBytes =
        Array.create<byte> (image.Width * image.Height * Unsafe.SizeOf<Rgba32>()) 0uy

    image.CopyPixelDataTo(pixelBytes)
    MD5.HashData(pixelBytes) |> Convert.ToBase64String

let private formatDownloadError comicName (exn: exn) =
    let message = $"Comic download failed for {comicName}: {exn.Message}"

    if message.Length > 4000 then
        message.Substring(0, 4000)
    else
        message

let downloadImages (logger: ILogger) (storageContext: StorageContext.Context) =
    let httpClient = new HttpClient()
    
    logger.LogInformation "Loading settings"

    let oneDriveSettings =
        BlobClient.load<OneDrive.OneDriveSettings> "settings" "onedrive" storageContext

    let telegramSettings =
        BlobClient.load<Telegram.TelegramSettings> "settings" "telegram" storageContext

    logger.LogInformation(
        "Settings loaded. OneDrive configured: {HasOneDriveSettings}. Telegram configured: {HasTelegramSettings}.",
        Option.isSome oneDriveSettings,
        Option.isSome telegramSettings
    )

    let downloadComic comicName (downloadFunction: Task<byte array option * string option>) =
        task {
            try
                logger.LogInformation("Starting comic download for {ComicName}", [| box comicName |])
                let! (bytes, caption) = downloadFunction

                match bytes with
                | None ->
                    logger.LogInformation("Skipping {ComicName} because the source returned no image", [| box comicName |])
                | Some bytes ->

                    logger.LogInformation("Downloaded {ByteCount} bytes for {ComicName}", [| box bytes.Length; box comicName |])
                    let fileName = getFileName bytes
                    let hash = getHash comicName bytes

                    if BlobClient.exists comicName hash storageContext then
                        logger.LogInformation("Skipping {ComicName} because hash {Hash} already exists", [| box comicName; box hash |])
                        ()
                    else
                        logger.LogInformation("New comic detected for {ComicName}; writing filename {FileName} with hash {Hash}", [| box comicName; box fileName; box hash |])

                        let! oneDriveResult = OneDrive.uploadFile logger oneDriveSettings comicName fileName bytes

                        match oneDriveResult with
                        | Ok _ -> logger.LogInformation("OneDrive upload succeeded for {ComicName}", [| box comicName |])
                        | Error error -> logger.LogWarning(error, "OneDrive upload returned an error for {ComicName}", [| box comicName |])

                        let! telegramResult = Telegram.sendMessage logger telegramSettings bytes caption

                        match telegramResult with
                        | Ok _ -> logger.LogInformation("Telegram image send succeeded for {ComicName}", [| box comicName |])
                        | Error error -> logger.LogWarning(error, "Telegram image send returned an error for {ComicName}", [| box comicName |])

                        BlobClient.storeBinary comicName hash bytes storageContext |> ignore
                        logger.LogInformation("Stored {ComicName} bytes in object storage under hash {Hash}", [| box comicName; box hash |])
                        ()
            with exn ->
                logger.LogError(exn, "Could not download {ComicName}", comicName)

                let! _ =
                    Telegram.sendText logger telegramSettings (formatDownloadError comicName exn)

                ()
        }

    task {
        logger.LogInformation("Starting comic download batch")
        do! downloadComic "Lunch TU" (downloadTU httpClient "lunch")
        do! downloadComic "Dunce" (downloadTU httpClient "dunce")

        do! downloadComic "XKCD" (downloadXkcd httpClient)

        do! downloadComic "Lunch E24" (downloadE24 httpClient)

        do! downloadComic "Calvin and Hobbes" (downloadCalvinAndHobbes storageContext)

        do! downloadComic "The Far Side" (downloadFarSide httpClient)

        do! downloadComic "Swords" (downloadRss httpClient "https://swordscomic.com/comic/feed/")
        
        logger.LogInformation("Comic download batch completed")
    }

File: download.go
package main

import (
	"fmt"
	"log/slog"
)

type comicJob struct {
	name     string
	download func() (*ComicResult, error)
}

func RunDownload(app *AppContext, logger *slog.Logger) error {
	logger.Info("Loading settings")

	oneDriveSettings, err := app.GetOneDriveSettings()
	if err != nil {
		logger.Warn("Could not load OneDrive settings", "error", err)
	}
	telegramSettings, err := app.GetTelegramSettings()
	if err != nil {
		logger.Warn("Could not load Telegram settings", "error", err)
	}

	logger.Info("Settings loaded",
		"hasOneDrive", oneDriveSettings != nil,
		"hasTelegram", telegramSettings != nil)

	var tgClient *TelegramClient
	if telegramSettings != nil {
		tgClient, err = NewTelegramClient(telegramSettings)
		if err != nil {
			logger.Warn("Could not create Telegram client", "error", err)
		}
	}

	comics := []comicJob{
		{"Lunch TU", DownloadLunchTU},
		{"XKCD", DownloadXKCD},
		{"Lunch E24", DownloadLunchE24},
		{"The Far Side", DownloadFarSide},
		{"Swords", func() (*ComicResult, error) {
			return DownloadRSS("https://swordscomic.com/comic/feed/")
		}},
	}

	logger.Info("Starting comic download batch")
	for _, comic := range comics {
		processComic(app, logger, oneDriveSettings, tgClient, comic)
	}
	logger.Info("Comic download batch completed")
	return nil
}

func processComic(app *AppContext, logger *slog.Logger, oneDrive *OneDriveSettings, tg *TelegramClient, job comicJob) {
	logger.Info("Starting comic download", "comic", job.name)

	result, err := job.download()
	if err != nil {
		logger.Error("Could not download comic", "comic", job.name, "error", err)
		sendErrorToTelegram(logger, tg, job.name, err)
		return
	}
	logger.Info("Downloaded comic", "comic", job.name, "bytes", len(result.Data))

	hash, err := hashImagePixels(result.Data)
	if err != nil {
		logger.Error("Could not hash comic", "comic", job.name, "error", err)
		sendErrorToTelegram(logger, tg, job.name, err)
		return
	}

	exists, err := app.BlobExists(job.name, hash)
	if err != nil {
		logger.Error("BlobExists check failed", "comic", job.name, "error", err)
		sendErrorToTelegram(logger, tg, job.name, err)
		return
	}
	if exists {
		logger.Info("Skipping duplicate comic", "comic", job.name, "hash", hash)
		return
	}

	logger.Info("New comic detected", "comic", job.name, "hash", hash)
	fileName := getFileName(result.Data)

	if oneDrive != nil {
		accessToken, _, odErr := RefreshToken(oneDrive.ClientID, oneDrive.RefreshToken)
		if odErr != nil {
			logger.Warn("OneDrive token refresh failed", "comic", job.name, "error", odErr)
		} else if odErr = UploadToOneDrive(accessToken, job.name, fileName, result.Data); odErr != nil {
			logger.Warn("OneDrive upload failed", "comic", job.name, "error", odErr)
		} else {
			logger.Info("OneDrive upload succeeded", "comic", job.name)
		}
	} else {
		logger.Warn("Skipping OneDrive (no settings)", "comic", job.name)
	}

	if tg != nil {
		caption := normalizeCaption(captionStr(result.Caption))
		if tgErr := tg.SendPhoto(result.Data, fileName, caption); tgErr != nil {
			logger.Warn("Telegram send failed", "comic", job.name, "error", tgErr)
		} else {
			logger.Info("Telegram send succeeded", "comic", job.name)
		}
	} else {
		logger.Warn("Skipping Telegram (no settings)", "comic", job.name)
	}

	ct := detectContentType(result.Data)
	if err := app.WriteBlob(job.name, hash, ct, result.Data); err != nil {
		logger.Error("Could not store comic in GCS", "comic", job.name, "error", err)
		return
	}
	logger.Info("Stored comic in GCS", "comic", job.name, "hash", hash)
}

func captionStr(caption *string) string {
	if caption == nil {
		return ""
	}
	return *caption
}

func sendErrorToTelegram(logger *slog.Logger, tg *TelegramClient, comicName string, err error) {
	if tg == nil {
		return
	}
	msg := fmt.Sprintf("Comic download failed for %s: %s", comicName, err.Error())
	if len(msg) > 4000 {
		msg = msg[:4000]
	}
	if tgErr := tg.SendText(msg); tgErr != nil {
		logger.Warn("Could not send Telegram error message", "error", tgErr)
	}
}

func RunRefreshOneDriveToken(app *AppContext, logger *slog.Logger) error {
	logger.Info("Starting OneDrive token refresh")

	settings, err := app.GetOneDriveSettings()
	if err != nil {
		return fmt.Errorf("load OneDrive settings: %w", err)
	}
	if settings == nil {
		logger.Warn("Skipping OneDrive refresh (no settings)")
		return nil
	}

	logger.Info("Loaded OneDrive settings")
	_, newRefresh, err := RefreshToken(settings.ClientID, settings.RefreshToken)
	if err != nil {
		return fmt.Errorf("refresh token: %w", err)
	}

	settings.RefreshToken = newRefresh
	if err := app.PutOneDriveSettings(settings); err != nil {
		return fmt.Errorf("store refreshed token: %w", err)
	}
	logger.Info("Stored refreshed OneDrive token")
	return nil
}
